lib/spack/spack/compiler.py

```diff
@@ -270,8 +271,8 @@ def opt_flags(self):
     PrgEnv_compiler = None
 
     def __init__(self, cspec, operating_system, target,
-                 paths, modules=None, alias=None, environment=None,
-                 extra_rpaths=None, enable_implicit_rpaths=None,
+                 paths, modules=[], alias=None, environment={},
+                 extra_rpaths=[], enable_implicit_rpaths=None,
```

Comment:
Don't do this.  Set it to `None` here and use:

```python
    self.varname = varname or {}
```

in the code.  The reason you want to do this is because the default values are evaluated at function definition time.  So, `{}` or `[]` as a default value ends up being the same `list` or `dict` for every invocation of `__init__()`, and therefore for every instance of this.  If anyone changes `environment`, they'll inadvertently change it for every instance.

As a rule of thumb, I just always use `None` when I need a default value that's not a constant or immutable type.